src/NuGet.Core/NuGet.Protocol/PackagesFolder/LocalPackageFileCache.cs

```diff
@@ -63,7 +63,19 @@ public Lazy<IReadOnlyList<string>> GetOrAddFiles(string expandedPath)
         public Lazy<string> GetOrAddSha512(string sha512Path)
         {
             return _sha512Cache.GetOrAdd(sha512Path,
-                e => new Lazy<string>(() => File.ReadAllText(e)));
+                e => new Lazy<string>(() =>
+                {
+                    try
+                    {
+                        return File.ReadAllText(e);
+                    }
+                    catch (Exception)
+                    {
+                        var ignore = new Lazy<string>();
+                        _sha512Cache.TryRemove(e, out ignore);
```

Comment:
This feels a little hacky to remove as part of the delegate for the add. The caching around File.Exists is for the exact reason. My guess is that your perf numbers are from before this optimization was made where each package was checked for each projects, the code now caches all disk operations between projects.

If this does have a good perf benefit despite being ugly, I'm fine with it.

Reply:
Discussed offline, we'll rather have `FileExists` check and remove `FileRead` for restore no-op